File: android/PriceFinder/app/src/main/java/enterprises/stimes/pricefinder/models/HomeItem.java
package enterprises.stimes.pricefinder.models;

import java.util.ArrayList;

/**
 * Created by Jacob on 6/4/2016.
 */
public class HomeItem {

    public long groupId;
    public String title;
    public double targetPrice;
    public double lowestPrice;
    public boolean foundTarget;
    public ArrayList<Item> items = new ArrayList<Item>();

    //date added

    public HomeItem(long groupId, String title, double target, Item[] items){
        this.groupId = groupId;
        this.title = title;
        this.targetPrice = target;

        for(Item item : items){
            this.items.add(item);
        }

        findLowestAndTarget();
    }

    public HomeItem(long groupId, String title, double target, ArrayList<Item> items){
        this.groupId = groupId;
        this.title = title;
        this.targetPrice = target;
        this.items.addAll(items);

        findLowestAndTarget();
    }

    private void findLowestAndTarget(){
        lowestPrice = 1000000.0;
        for(int i=0; i<items.size(); i++){
            Item itemI = items.get(i);
            if(itemI.price < lowestPrice){
                lowestPrice = itemI.price;
            }
        }
        if(lowestPrice < targetPrice){
            foundTarget = true;
        }
    }
}

File: ios/PriceFinder/PriceFinder/Models/HomeItem.swift
//
//  HomeItem.swift
//  PriceFinder
//
//  Created by jacob stimes on 6/7/16.
//  Copyright © 2016 stimes.enterprises. All rights reserved.
//

import Foundation

/*!
	@brief Represents a user's saved product search. Not a single item, but can contain multiple items related to title - the user's search query.
	I didn't want to use the managed object counterpart to avoid unintentional database updates
*/
class HomeItem {
    
    var groupId: Int16?
    
    var title: String!
    
    var targetPrice: Double!
    
    var lowestPrice: Double!
    
    var items: [Item] = []
    
    var dateAdded: NSDate?
    
    init(groupId: Int16, title:String, targetPrice price:Double, dateAdded: NSDate, withItems items:[Item]){
        self.groupId = groupId;
        self.title = title;
        self.targetPrice = price;
        self.dateAdded = dateAdded;
        self.sortAndAddItems(items);
    }
    
    init(title:String, targetPrice price:Double, withItems items:[Item]){
        self.title = title;
        self.targetPrice = price;
        self.sortAndAddItems(items);
    }
    
    private func sortAndAddItems(items: [Item]) {
        //Sort so that those with the largest value of (targetPrice - lowestPrice) are at top:
        // i.e. the biggest deals show up at top:
        self.items.appendContentsOf(items.sort( { (first: Item, second: Item ) -> Bool in
            let firstDiff = self.targetPrice - first.price;
            let secondDiff = self.targetPrice - second.price;
            
            return firstDiff > secondDiff;
        }));
        
        self.lowestPrice = self.items.first!.price;
    }
}
